platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/merkle/disk/blockstream/StateChangesObserverImpl.java

```diff
@@ -0,0 +1,278 @@
+import com.hedera.hapi.node.state.congestion.CongestionLevelStarts;
+import com.hedera.hapi.node.state.consensus.Topic;
+import com.hedera.hapi.node.state.contract.Bytecode;
+import com.hedera.hapi.node.state.contract.SlotKey;
+import com.hedera.hapi.node.state.contract.SlotValue;
+import com.hedera.hapi.node.state.file.File;
+import com.hedera.hapi.node.state.recordcache.TransactionRecordEntry;
+import com.hedera.hapi.node.state.schedule.Schedule;
+import com.hedera.hapi.node.state.schedule.ScheduleList;
+import com.hedera.hapi.node.state.throttles.ThrottleUsageSnapshots;
+import com.hedera.hapi.node.state.token.*;
+import com.hedera.hapi.node.transaction.ExchangeRateSet;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Observer watches a Record and a Stack. It ties the changes in a stack to the transactions that created the state
+ * changes. When a stack is committed we record all the changes that are a result of that commit.
+ *
+ * <p>What we'd like is a List of StateChanges that are tied to a SingleTransactionRecord, but right now this ties state
+ *    changes to a {@code stream<SingleTransactionRecord>} which is not ideal but will give us a good estimate of data.
+ *
+ * <p>To do that, we need an order of operations such that when a stack is committed, for a given transaction, we can let
+ *    the RecordListBuilder know what the changes are for that transaction. We don't want to include the changes for the
+ *    child transactions as those should be included in the RecordListBuilder for the child transactions.
+ *
+ * <p>Any time we add a new protobuf type to be serialized to state, we need to add it to the relevant methods below
+ *    that map the protobuf type to the OneOf type. It would be desirable if in the future PBJ generated these methods.
+ *    That would ensure we would only need to add the type to the protobuf definition.
+ */
+public class StateChangesObserverImpl implements StateChangesObserver {
+    private List<StateChange> stateChanges;
+    private LinkedList<StateChange> endOfRoundStateChanges;
+    private HashMap<String, StateChange> endOfRoundSingletonStateChanges;
+
+    /**
+     * Implementation of the StateChangesObserver for Block Streams. This should only ever be instantiated by
+     * SingletonStateChangesObserver.
+     */
+    protected StateChangesObserverImpl() {
+        resetStateChanges();
+        resetEndOfRoundStateChanges();
+    }
+
+    public void resetStateChanges() {
+        stateChanges = new ArrayList<>();
+    }
+
+    public void resetEndOfRoundStateChanges() {
+        endOfRoundStateChanges = new LinkedList<>();
+        endOfRoundSingletonStateChanges = new HashMap<>();
+    }
+
+    public boolean hasRecordedStateChanges() {
+        return !stateChanges.isEmpty();
```

Comment:
this will throw a NPE if reset has not called before. It would be way better to define stateChanges as final and clear it in resetStateChanges